File: experimental/zig/src/dynamicprocesses.zig
// Actress Copyright (C) 2024  Bjørn Tore Svinningen
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

const std = @import("std");
const Io = std.Io;

const actress = @import("actress.zig");
const slog = @import("slog.zig");
const Context = @import("context.zig").Context;
const Event = @import("events.zig").Event;
const EventName = @import("events.zig").EventName;
const ETFunc = @import("events.zig").ETFunc;
const Process = actress.Process;
const setPreviousEvent = @import("staticprocesses.zig").setPreviousEvent;

// Holds information about what process functions who belongs to what
// event, and also a map of the started processes.
pub const dynamicProcesses = struct {
    procMap: std.StringHashMap(*Process),
    mu: Io.Mutex = .init,

    // Add a new Event and it's process to the processes map.
    //
    // *******************************************************************
    // TODO: Consider if we still need this function. It is not in use.
    // *******************************************************************
    pub fn Add(p: *dynamicProcesses, io: Io, et: EventName, proc: *Process) !void {
        // Check if a process for the same event is defined, and if so we
        // cancel the current process before we replace it with a new one.
        p.mu.lockUncancelable(io);
        defer p.mu.unlock(io);
        if (p.procMap.get(et)) |existing| {
            existing.Cancel();
        }
        try p.procMap.put(et, proc);
    }

    // Delete an Event and it's process from the processes map.
    pub fn Delete(p: *dynamicProcesses, io: Io, et: EventName) void {
        p.mu.lockUncancelable(io);
        defer p.mu.unlock(io);
        // p.procMap.get(et).?.Cancel();
        _ = p.procMap.remove(et);
        std.debug.print("deleted process {s}\n", .{et});
    }

    // Checks if the event is defined in the processes map, and returns true if it is.
    pub fn IsEventDefined(p: *dynamicProcesses, io: Io, ev: EventName) bool {
        p.mu.lockUncancelable(io);
        defer p.mu.unlock(io);
        if (p.procMap.get(ev) == null) {
            return false;
        }

        return true;
    }
};

// Prepare and return a new *dynamicProcesses structure.
pub fn newDynamicProcesses(allocator: std.mem.Allocator) !*dynamicProcesses {
    const p = try allocator.create(dynamicProcesses);
    p.* = dynamicProcesses{
        .procMap = std.StringHashMap(*Process).init(allocator),
    };
    return p;
}

// Will create and return a new UUID prefix with "ED-".
// The caller owns the returned memory.
pub fn NewUUID(allocator: std.mem.Allocator, io: Io) ![]const u8 {
    // Generate a random version 4 UUID like the Go version does with
    // the google/uuid package.
    var b: [16]u8 = undefined;
    io.random(&b);
    b[6] = (b[6] & 0x0f) | 0x40;
    b[8] = (b[8] & 0x3f) | 0x80;

    const u = try std.fmt.allocPrint(allocator, "ED-{x:0>2}{x:0>2}{x:0>2}{x:0>2}-{x:0>2}{x:0>2}-{x:0>2}{x:0>2}-{x:0>2}{x:0>2}-{x:0>2}{x:0>2}{x:0>2}{x:0>2}{x:0>2}{x:0>2}", .{ b[0], b[1], b[2], b[3], b[4], b[5], b[6], b[7], b[8], b[9], b[10], b[11], b[12], b[13], b[14], b[15] });
    return u;
}

// ------------------------------------------------------------------------------
// Events and event functions.
// ------------------------------------------------------------------------------

// Router for normal events.
pub const EDRouter: EventName = "EDRouter";

// Process function for routing and handling events. Will check
// and route the event to the correct process.
pub const edRouterFn = ETFunc{ .run = edRouterFnRun };

fn edRouterFnRun(userdata: ?*anyopaque, ctx: *Context, p: *Process) void {
    _ = userdata;

    defer {
        slog.Debug("edRouterFn", .{ "stopping process with defered stop", "" });
        p.Stop();
    }

    while (true) {
        var ev = p.DynamicEventCh.getOne(p.io) catch {
            slog.Debug("edRouterFn", .{ "got ctx.Done, on", p.Config.NodeName });
            return;
        };

        // If there is a next event defined, we make a copy of all the fields  of the current event,
        // and put that as the previousEvent on the next event. We can use this information later
        // if need to check something in the previous event.
        if (ev.NextEvent) |ne| {
            // Keep the information about the current event, so we are able to check for things
            // like ackTimeout and what node to reply back to if ack should be given.
            setPreviousEvent(p.allocator, ne, &ev);
        }

        // Dynamic processes can take a little longer to start up and be
        // registered in the map. We check here if process is registred,
        // and if it is not we retry.
        // The checking is done in a task so the router don't block
        // here waiting, and we continue with the next event in the queue.
        //
        // NB, Zig specific: the inflight counter of the process is
        // incremented while the map lock is held, so a concurrent Stop of
        // the process waits with freeing it's memory until the delivery
        // below is done. The Go version needs no such pin, the garbage
        // collector keeps the process alive through the pointer.
        p.DynamicProcesses.mu.lockUncancelable(p.io);
        const procOpt = p.DynamicProcesses.procMap.get(ev.Name);
        if (procOpt) |proc| {
            _ = proc.inflight.fetchAdd(1, .acq_rel);
        }
        p.DynamicProcesses.mu.unlock(p.io);

        const proc = procOpt orelse {
            ctx.spawn(edRouterRetryDeliver, .{ p, ev }) catch {
                ev.deinit(p.allocator);
                return;
            };

            // The above task will wait, and check if the process becomes
            // available, and send the event if it the process eventually found,
            // so we can continue with the next event in the queue.
            continue;
        };
        defer _ = proc.inflight.fetchSub(1, .release);

        // Process was registered. Deliver the event to the process InCh.
        slog.Debug("edRouterFn", .{ "on", p.Config.NodeName, "Routing event", p.Event, "node", p.Config.NodeName, "name", ev.Name, "Inch", proc.InCh });

        proc.InCh.putOne(p.io, ev) catch |err| switch (err) {
            error.Canceled => {
                ev.deinit(p.allocator);
                return;
            },
            error.Closed => {
                ev.deinit(p.allocator);
                continue;
            },
        };
    }
}

// The retry task started by the router when no process was registered for
// the event yet. The equivalent of the `go func(ev Event)` in the Go version.
fn edRouterRetryDeliver(p: *Process, evIn: Event) void {
    const ev = evIn;

    // Try to 3 times to deliver the message.
    var i: usize = 0;
    while (i < 3) : (i += 1) {
        // The inflight counter is incremented while the map lock is held,
        // so a concurrent Stop of the found process waits with freeing it
        // until the delivery is done. See the comment in edRouterFnRun.
        p.DynamicProcesses.mu.lockUncancelable(p.io);
        const procOpt = p.DynamicProcesses.procMap.get(ev.Name);
        if (procOpt) |proc| {
            _ = proc.inflight.fetchAdd(1, .acq_rel);
        }
        p.DynamicProcesses.mu.unlock(p.io);

        const proc = procOpt orelse {
            slog.Error("edRouterFn", .{ "on", p.Config.NodeName, "found no process registered for the event type", ev.Name, "ev.DstNode", ev.DstNode });
            p.io.sleep(.fromSeconds(1), .awake) catch {
                ev.deinit(p.allocator);
                return;
            };
            continue;
        };
        defer _ = proc.inflight.fetchSub(1, .release);

        // Process is now registred, so we can safely put
        //the event on the InCh of the process.
        proc.InCh.putOne(p.io, ev) catch {
            ev.deinit(p.allocator);
            return;
        };

        return;
    }

    // All retries failed, the event is dropped and freed.
    ev.deinit(p.allocator);
}

// EDSync is used to syncronize events.
// The EDSyncFn that is to be used with this event type takes a
// signal channel, we can then use this event type to signal that
// another event is done before we continue by setting this event
// type as the NextEvent.
pub const EDSync: EventName = "EDSync";

// EtSyncFn is the function that will be used to syncronize events.
// It takes a channel that will be used to send a signal on when the
// EDSync event is executed.
//
// In general EDSync is used to syncronize one-off events, so delete
// the process after it is done with it's sync job.
//
// The Go version uses a chan struct{}, the Zig version uses a queue of
// u8 since a queue element can not be zero sized.
pub fn EDSyncFn(syncCh: *Io.Queue(u8)) ETFunc {
    return ETFunc{ .userdata = syncCh, .run = edSyncFnRun };
}

fn edSyncFnRun(userdata: ?*anyopaque, ctx: *Context, p: *Process) void {
    _ = ctx;
    const syncCh: *Io.Queue(u8) = @ptrCast(@alignCast(userdata.?));

    while (true) {
        p.SignalReady();

        const ev = p.InCh.getOne(p.io) catch return;
        defer ev.deinit(p.allocator);

        syncCh.putOne(p.io, 0) catch return;
    }
}

File: experimental/rust/src/dynamicprocesses.rs
// Actress Copyright (C) 2024  Bjørn Tore Svinningen
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.

use crate::actress::{Process, ProcFn, ETFunc};
use crate::context::Context;
use crate::events::EventName;
use crate::staticprocesses::CopyEventFields;

use crossbeam_channel::select;
use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use std::time::Duration;

// Holds information about what process functions belong to what event, and a
// map of the started processes. The purpose of dynamic processes is to have
// short lived processes that can be quickly started and removed again when
// their job is done. The only difference from a static process is that dynamic
// processes can be deleted from the map.
pub struct dynamicProcesses {
    pub procMap: Mutex<HashMap<EventName, Arc<Process>>>,
}

impl dynamicProcesses {
    // Add a new Event and its process to the processes map. Cancels any
    // existing process for the same event first.
    //
    // NB: This mirrors the Go method that the comment marks as currently
    // unused; kept for parity.
    pub fn Add(&self, et: EventName, proc: Arc<Process>) {
        let mut m = self.procMap.lock().unwrap();
        if let Some(old) = m.get(&et) {
            (old.Cancel)();
        }
        m.insert(et, proc);
    }

    // Delete an Event and its process from the processes map.
    pub fn Delete(&self, et: &EventName) {
        self.procMap.lock().unwrap().remove(et);
        println!("deleted process {}", et);
    }

    // Checks if the event is defined in the processes map.
    pub fn IsEventDefined(&self, ev: &EventName) -> bool {
        self.procMap.lock().unwrap().contains_key(ev)
    }
}

// Prepare and return a new dynamicProcesses structure.
pub fn newDynamicProcesses() -> dynamicProcesses {
    dynamicProcesses {
        procMap: Mutex::new(HashMap::new()),
    }
}

// Will create and return a new UUID prefixed with "ED-".
pub fn NewUUID() -> String {
    format!("ED-{}", uuid::Uuid::new_v4())
}

// ------------------------------------------------------------------------------
// Events and event functions.
// ------------------------------------------------------------------------------

// Router for dynamic events.
pub const EDRouter: EventName = EventName::new("EDRouter");

// Process function for routing and handling dynamic events.
pub fn edRouterFn(_ctx: Context, p: Arc<Process>) -> ProcFn {
    Box::new(move || {
        loop {
            select! {
                recv(p.DynamicEventCh.rx) -> msg => {
                    let mut ev = match msg { Ok(e) => e, Err(_) => break };

                    // If there is a next event defined, copy the current event
                    // fields into the next event's PreviousEvent.
                    if ev.NextEvent.is_some() {
                        let prev = CopyEventFields(&ev);
                        if let Some(ne) = ev.NextEvent.as_mut() {
                            ne.PreviousEvent = Some(Box::new(prev));
                        }
                    }

                    // Dynamic processes can take a little longer to start up and
                    // be registered in the map. If the process is not registered
                    // yet, we retry in a separate thread so the router does not
                    // block, and continue with the next event in the queue.
                    let registered = p.DynamicProcesses.procMap.lock().unwrap().contains_key(&ev.Name);

                    if !registered {
                        let p2 = p.clone();
                        std::thread::spawn(move || {
                            // Try 3 times to deliver the message.
                            for _ in 0..3 {
                                let in_tx = {
                                    let m = p2.DynamicProcesses.procMap.lock().unwrap();
                                    m.get(&ev.Name).map(|proc| proc.InCh.tx.clone())
                                };

                                match in_tx {
                                    None => {
                                        crate::log_error!(
                                            "edRouterFn: on {}: found no process registered for the event type {}, ev.DstNode {}",
                                            p2.Config.NodeName, ev.Name, ev.DstNode
                                        );
                                        std::thread::sleep(Duration::from_secs(1));
                                        continue;
                                    }
                                    Some(tx) => {
                                        let _ = tx.send(ev);
                                        return;
                                    }
                                }
                            }
                        });
                        continue;
                    }

                    // Process was registered. Deliver the event to its InCh.
                    let in_tx = {
                        let m = p.DynamicProcesses.procMap.lock().unwrap();
                        m.get(&ev.Name).map(|proc| proc.InCh.tx.clone())
                    };
                    if let Some(tx) = in_tx {
                        let _ = tx.send(ev);
                    }
                }
                recv(p.Ctx.done()) -> _ => break,
            }
        }

        p.Stop();
    })
}

// EDSync is used to synchronize events. The EDSyncFn takes a signal channel; we
// can then use this event type to signal that another event is done before we
// continue, by setting EDSync as the NextEvent.
pub const EDSync: EventName = EventName::new("EDSync");

// EDSyncFn is the function used to synchronize events. It takes a channel that
// will be used to send a signal when the EDSync event is executed.
//
// In general EDSync is used to synchronize one-off events, so delete the
// process after it is done with its sync job.
pub fn EDSyncFn(syncCh: crossbeam_channel::Sender<()>) -> ETFunc {
    Box::new(move |ctx: Context, p: Arc<Process>| -> ProcFn {
        Box::new(move || {
            loop {
                p.SignalReady();

                select! {
                    recv(p.InCh.rx) -> msg => {
                        if msg.is_err() { return; }
                        let _ = syncCh.send(());
                    }
                    recv(ctx.done()) -> _ => return,
                }
            }
        })
    })
}
